lib/Sema/DerivedConformanceComparable.cpp

```diff
@@ -0,0 +1,411 @@
+                     /*FuncLoc=*/SourceLoc(), name, /*NameLoc=*/SourceLoc(),
+                     /*Throws=*/false, /*ThrowsLoc=*/SourceLoc(),
+                     /*GenericParams=*/nullptr,
+                     params,
+                     TypeLoc::withoutLoc(boolTy),
+                     parentDC);
+  comparableDecl->setImplicit();
+  comparableDecl->setUserAccessible(false);
+
+  // Add the @_implements(Comparable, < (_:_:)) attribute
+  if (generatedIdentifier != C.Id_LessThanOperator) {
+    auto comparable = C.getProtocol(KnownProtocolKind::Comparable);
+    auto comparableType = comparable->getDeclaredType();
+    auto comparableTypeLoc = TypeLoc::withoutLoc(comparableType);
+    SmallVector<Identifier, 2> argumentLabels = { Identifier(), Identifier() };
+    auto comparableDeclName = DeclName(C, DeclBaseName(C.Id_LessThanOperator),
+                                   argumentLabels);
+    comparableDecl->getAttrs().add(new (C) ImplementsAttr(SourceLoc(),
+                                                          SourceRange(),
+                                                          comparableTypeLoc,
+                                                          comparableDeclName,
+                                                          DeclNameLoc()));
+  }
+
+  if (!C.getLessThanIntDecl()) {
+    derived.ConformanceDecl->diagnose(diag::no_less_than_overload_for_int);
+    return nullptr;
+  }
+
+  comparableDecl->setBodySynthesizer(bodySynthesizer);
+
+  comparableDecl->copyFormalAccessFrom(derived.Nominal, /*sourceIsParentContext*/ true);
+
+  // Add the operator to the parent scope.
+  derived.addMembersToConformanceContext({comparableDecl});
+
+  return comparableDecl;
+}
+
+bool 
+DerivedConformance::canDeriveComparable(DeclContext *context, NominalTypeDecl *declaration) {
+  // The type must be an enum.
+  if (EnumDecl *const enumeration = dyn_cast<EnumDecl>(declaration)) {
+    // The cases must not have non-comparable associated values or raw backing
+    auto comparable = context->getASTContext().getProtocol(KnownProtocolKind::Comparable);
+    return allAssociatedValuesConformToProtocol(context, enumeration, comparable) && !enumeration->hasRawType();
+  } else {
+    return false;
+  }
+}
+
+ValueDecl *DerivedConformance::deriveComparable(ValueDecl *requirement) {
+  if (checkAndDiagnoseDisallowedContext(requirement))
+    return nullptr;
+  // Build the necessary decl.
+  if (requirement->getBaseName() == "<") {
+    if (EnumDecl const *const enumeration = dyn_cast<EnumDecl>(this->Nominal)) {
+      auto bodySynthesizer = !enumeration->hasCases() 
+        ? &deriveBodyComparable_enum_uninhabited_lt 
+        : enumeration->hasOnlyCasesWithoutAssociatedValues()
+            ? &deriveBodyComparable_enum_noAssociatedValues_lt
+            : &deriveBodyComparable_enum_hasAssociatedValues_lt; 
+      return deriveComparable_lt(*this, bodySynthesizer);
+    } else {
+      llvm_unreachable("todo");
```

Comment:
Uh oh... Does returning `nullptr` still do the right thing here?

Reply:
if not, then the existing `DerivedConformance::deriveEquatable` function also has this problem

Reply:
Hm, that unreachable is from 2014.  I apologize for giving you refactoring tasks for old code you didn't write, but I think if we do a better job in this file we can propagate it back to the others.

Reply:
what should it be doing instead of returning `nullptr`?

Reply:
Let me be more explicit: This branch is really unreachable, so it should be an assert.

Reply:
done